### ISSUE TITLE
Some DateTimeTest format tests fail on the upcoming Fedora 39

### ISSUE THREAD

**Issue by tmds** [dotnet/runtime#84763]
```
System.Tests.DateTimeTests.TryFormat_MatchesToString(format: \"f\")
System.Tests.DateTimeTests.TryFormat_MatchesToString(format: \"F\")
System.Tests.DateTimeTests.TryFormat_MatchesToString(format: \"g\")
System.Tests.DateTimeTests.TryFormat_MatchesToString(format: \"G\")
System.Tests.DateTimeTests.TryFormat_MatchesToString(format: \"t\")
System.Tests.DateTimeTests.TryFormat_MatchesToString(format: \"T\")
System.Tests.DateTimeTests.TryFormat_MatchesToString(format: \"U\")
System.Tests.DateTimeTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"f\")
System.Tests.DateTimeTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"F\")
System.Tests.DateTimeTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"g\")
System.Tests.DateTimeTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"G\")
System.Tests.DateTimeTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"t\")
System.Tests.DateTimeTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"T\")
System.Tests.DateTimeTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"U\")
System.Tests.DateTimeOffsetTests.TryFormat_ToString_EqualResults
System.Tests.DateTimeOffsetTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"f\")
System.Tests.DateTimeOffsetTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"F\")
System.Tests.DateTimeOffsetTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"g\")
System.Tests.DateTimeOffsetTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"G\")
System.Tests.DateTimeOffsetTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"t\")
System.Tests.DateTimeOffsetTests.ParseExact_ToStringThenParseExactRoundtrip_Success(standardFormat: \"T\")
```

It's likely there was a regression in https://github.com/dotnet/runtime/pull/83589 which was not caught in CI.

The test failures are from our CI machine. I will reproduce the issue locally and share more information.

cc @tarekgh @omajid

**Comment by tarekgh** [comment]
The problem is in DateTime.TryFormat with `Utf-8`. This assumes all date/time pattern characters are ascii which is not the case as CLDR started to use `U+202F` as the space character. The following line should convert the character to Utf-8 and store it to the buffer instead of just casting it. 

https://github.com/dotnet/runtime/blob/9aefa9daa141bb7d9ba3f2b373d4b050c9b243fe/src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormat.cs#L730

**Comment by tmds** [comment]
I had tried to reproduce the issue in a Fedora 38, and a Fedora 39 container yesterday and to my surprise it didn't reproduce.

> The problem is in DateTime.TryFormat with Utf-8. This assumes all date/time pattern characters are ascii which is not the case as CLDR started to use U+202F as the space character. The following line should convert the character to Utf-8 and store it to the buffer instead of just casting it.

Thanks for pointing me to the problem.
This matches with the failures: they show `/` (0x2F) which is the lower byte of `U+202F`:

```
Expected: ···ril 18, 2023 5:10:51 AM
Actual:   ···ril 18, 2023 5:10:51/AM
```

**Comment by tarekgh** [comment]
> I had tried to reproduce the issue in a Fedora 38, and a Fedora 39 container yesterday and to my surprise it didn't reproduce.

I am able to produce the issue on edora 39. I just used `Fedora:39` docker image

**Comment by tmds** [comment]
> I am able to produce the issue on Fedora 39. I just used Fedora:39 docker image

Let me try again.

I'm using the image from registry.fedoraproject.org/fedora , which has a sha ae8f1aaa5e2b for Fedora 39.
Are we using the same one?

**Comment by tarekgh** [comment]
I just did `docker pull fedora:39` and then manually added all needed dependencies to build the runtime repo.

**Comment by tarekgh** [comment]
Maybe you only need to ensure using the latest ICU version in your container 

```
[root@b6b1f50cc61b tests]# dnf list installed | grep -i icu
libicu.x86_64                              72.1-2.fc38                 @rawhide
libicu-devel.x86_64                        72.1-2.fc38                 @rawhide
```

**Comment by tmds** [comment]
> Maybe you only need to ensure using the latest ICU version in your container

I have the same versions for icu.

And, the tests don't fail with the Docker Hub image for me either. :thinking:

I'm executing the tests by running `dotnet test` in `src/libraries/System.Runtime/tests`.

My env looks like:
```
# env
HOSTNAME=a876272645f7
DISTTAG=f39container
PWD=/root/runtime/src/libraries/System.Runtime/tests
FBR=f39
container=podman
HOME=/root
LANG=C.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=01;37;41:su=37;41:sg=30;43:ca=00:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.avif=01;35:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=01;36:*.au=01;36:*.flac=01;36:*.m4a=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.oga=01;36:*.opus=01;36:*.spx=01;36:*.xspf=01;36:*~=00;90:*#=00;90:*.bak=00;90:*.old=00;90:*.orig=00;90:*.part=00;90:*.rej=00;90:*.swp=00;90:*.tmp=00;90:*.dpkg-dist=00;90:*.dpkg-old=00;90:*.ucf-dist=00;90:*.ucf-new=00;90:*.ucf-old=00;90:*.rpmnew=00;90:*.rpmorig=00;90:*.rpmsave=00;90:
FGC=f39
TERM=xterm
SHLVL=1
PATH=/root/runtime/.dotnet:/root/.local/bin:/root/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
OLDPWD=/root/runtime
_=/usr/bin/env
```

My timezone is UTC:
```
# date +%Z
UTC
```

**Comment by tmds** [comment]
Aha, this does it:
```
export LANG=en_US.UTF-8
```

**Comment by tarekgh** [comment]
right, don't keep `C` locale as default. We use Invariant culture at that time which will not have `U+202F` in the time formats.

**Comment by tarekgh** [comment]
@tmds are you willing to open a PR for it? If not, I'll take care with it.

**Comment by tmds** [comment]
I was working on this patch: https://github.com/dotnet/runtime/compare/main...tmds:runtime:parse_nbsp.

This fixes the `FormatExceptions` I had mentioned in https://github.com/dotnet/runtime/pull/84963#issuecomment-1512746458.

If this looks like the way to go, I'll open up a PR.

**Comment by tmds** [comment]
Reopen to track fixing the FormatExceptions.
